Minor refactoring of InsertEdit.php

kamil-tekiela opened the pull request:
This is still minor refactoring despite many changes, as this class is in dire need of Value Objects and removal of globals, which I haven't touched in this PR. Static analysis and I are both restricted in the number of changes that can be made due to the fact that many of the values are typed as mixed. 

The goals of this refactoring:
- Removal of dead code
- Removal of redundant code
- Improving static analysis by providing type hints and type declarations
- Small refactorings to make code easier to read and spot bugs 

This PR does not:
- Change any functionality
- Add any new methods or classes
- Improve unit tests
- Change coding style

It seems that DbiDummy is broken (see the one test failure). The tests send an empty array where the `$result` object is expected, I fixed that by casting to an object. The mock methods seem to be returning an integer. I'm not sure what the correct behaviour of this should be, but it looks like it doesn't actually test the right functionality. Am I wrong?

This code contains ABI break, but I hope it won't be a problem. I changed some parameters of private and public methods. It will work the same but callers need to pass different arguments now. 

For the explanation of each change, see the description in commits.

kamil-tekiela committed:
Simplify condition

isset is not needed since the variable is set and we check explicitely for an array so it cannot be null either.
!empty is not needed because the variable is set, so we just need to check if it has non-empty value

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Reduce indentation thanks to early returns

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Collapse isset check

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant variable assignments

$specialCharsEncoded is immediately overwritten in if/else.
$noSupportTypes is always empty and never used in this code.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant !empty() checks on parameters and defined variables

if(!empty($var)) on defined variables is redundant and equivalent to if($var)

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant ternary

$dispval is declared to be a string so the ternary is redundant

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor showEmptyResultMessageOrSetUniqueCondition

Replace boolean variable with returns, which leads to redundant else
block. Remove redundant unset, which leads to redundant variable assignment.
The return statement can be replaced with a boolean cast.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant unset at the end of the method

When a variable leaves scope it is automatically unset

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
[RISKY] Remove temporary var and error supression

It's risky because I was not able to determine the purpose of the error
suppression. In case of error, the function seems to return -1, but that
has not been checked here either. It was possible to silence undefined
global variable, but in that case, why was it not done only on that one
line?

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant condition

The variable is declared as int|false so if it is not false and not 0,
the only logical choice is a non-zero integer, which is always true.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Misleading type hint leading to invalid unit tests

According to DbiMysqli the first argument should be mysqli_result.
The correct type hint cannot be array. It can be object though.
Tested with a debugger that the passed value is really mysqli_result.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Update psalm-baseline.xml

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove dead code

Remove $inputType. I left the comemnt for posterity, but the variable is
dead. If in the future someone decides to implement it again, then
introducing the variable will not be a problem.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Move block of code to where it should be

I am not sure about this block of code. I don't fully understand what
it is supposed to do. However, making the query at the start of the
function doesn't make sense to me.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Use array_keys instead of foreach with dummy variable

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Use is_null instead of isset

True conditions are easier to read than inverted conditions.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant variable and break

When the loop breaks then the variable is true.
When the loop finishes then the variable is false.
We can simply replace it with return true/false to make it simpler.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Add type declarations

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getEnumSetAndTimestampColumns

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant condition

According to searchColumnInForeigners $foreigner is dependant on values
from $foreigners which makes this check redundant.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getColumnEnumValues

Simplify parameter list, add psalm types, and simplify code

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getColumnSetValueAndSelectSize

Unfortunately, I cannot type hint the return value as then I would have
to type-hint the param as well.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove ternary operator

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getSelectOptionForUpload

Added types, removed NULL return (it is only outputted in Twig as
HTML string so NULL made no sense), and simpified parameters.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getMaxUploadSize

According to getMaxUploadSize the value can only be int.
Comparisons of numerical strings should be avoided. Psalm return type
can be added with the exact types.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getColumnSize

It takes a string as a second parameter now.
The reason for this is that during normal execution, the value comes
from column specifier e.g. VARCHAR(255), which is a string. The value
in spec_in_brackets could also represent ENUM/SET values, so it can't be
forced as an int.
This change required changing tests to pass a string instead of an array
with an int.
The behaviour should remain the same thanks to the (int) cast (previously
implicit, now explicit).

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove parameter $realNullValue

This parameter was always false. Instead define it as false inside
each method. The methods are private, so this should not cause any problems.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Add psalm-return

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Add an early return and clarify return type

This function should never return false

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Change return type to multidimensional array

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getCommentsMap

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Type hint $tableColumns as multidimensional array

This simplifies static analysis a little bit

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Update phpstan baseline

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Fix invalid unit test

The method getTableColumns should return a multidimensional array always.
Added proper type hint to avoid further mistakes and changed mock.
According to my analysis this method should always return a
multidimensional array because parameter $column to getColumns() is null.
Also, this test is pretty useless as it just seems to test the bahaviour
of array_values which behaves as an identity function right now.

Note: The DBAL needs urgent refactoring too. I imagine array_values is
redundant at the moment. The getColumns should only return a list or
columns or a single column. There's no need to reindex the list. I think
Psalm should have picked it up, but I didn't check.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant call to Util::extractColumnSpec()

Warning! There is no coverage in unit tests for this piece of code.
While I have analysed the code flow manually, it is possible the
duplicated call had some unknown to me purpose.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Adhere to the coding standard

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -690,26 +677,27 @@ private function getHtmlInput(
             $isUnsigned = $extractedColumnspec['unsigned'];
             $minMaxValues = $this->dbi->types->getIntegerRange($column['True_Type'], ! $isUnsigned);
-            $inputMinMax = 'min="' . $minMaxValues[0] . '" '
+            $inputMinMax = ' min="' . $minMaxValues[0] . '" '
```
Not sure if the space should be added there or where the variable is used

kamil-tekiela replied:
It doesn't matter really, but if it looks better then I will roll it back. In cases like this, I would usually use `implode()` rather than string concatenation but I didn't want to change too much here.

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -718,12 +706,12 @@ private function getHtmlInput(
-     * @param string $vkey   [multi_edit]['row_id']
-     * @param array  $column description of column in given table
+     * @param string $vkey      [multi_edit]['row_id']
+     * @param string $field_MD5 array index as an MD5 to avoid having special characters
```
```suggestion
     * @param string $fieldHashMd5 array index as an MD5 to avoid having special characters
```
This is maybe better for our coding standard

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -734,45 +722,46 @@ private function getSelectOptionForUpload($vkey, array $column)
         }
 
-        if (! empty($files)) {
+        if ($files) {
```
This is a bit vague, could it be more precise?

kamil-tekiela replied:
I can make it a more strict comparison. I didn't want to change the functionality, but the string comes from Template render so it probably wouldn't hurt to check strictly for an empty string rather than a non-empty-string (which includes string `'0'`). I doubt that the template render would ever return string `'0'` so it should be safe, and we can make it more clear with the help of early return.

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -899,17 +888,15 @@ private function getValueColumnForOtherDatatypes(
      * @return int field size
      */
-    private function getColumnSize(array $column, array $extractedColumnspec)
+    private function getColumnSize(array $column, string $spec_in_brackets): int
```
We use camel casing ;)

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -1119,21 +1103,21 @@ private function getSpecialCharsAndBackupFieldForExistingRow(
             $realNullValue = true;
             $data = '';
         } else {
+            $realNullValue = false;
```
Is it a bug fix?

kamil-tekiela replied:
I was not able to identify any buggy behaviour here. Technically it was possible to provide `true` here through the use of the parameter, but in all the uses I could find (including unit tests) it was always hardcoded to `false`. Was it a bug? I don't think so.

williamdes replied:
Okay, thanks for the fix then ;)

williamdes reviewed (changes requested):
Very nice work, thanks a lot! 
I did a bit of review from my mobile
Will have to have another look

kamil-tekiela committed:
Revert "Remove ternary operator"

This reverts commit 3cbec2cebb9efcc7d52988a5b5720c0e43dc6ddb.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Rename $field_MD5 to $fieldHashMd5

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Exclude string '0' from comparison and invert the order of operation

This helps with readability and it's unlikely that we would want to
check for string '0' here.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Rename $spec_in_brackets to $specInBrackets

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -1366,16 +1347,10 @@ public function executeSqlQuery(array $urlParams, array $query)
 
                 $insertId = $this->dbi->insertId();
-                if ($insertId !== false && $insertId != 0) {
+                if ($insertId) {
```
could it be more specific ?

kamil-tekiela replied:
Well... No. The value comes from `insertId()` that relies on undefined behaviour. It's unclear what the possible return values could be. The return value is type hinted as `int|false`, but I doubt that this is really the type of value that will be produced by this method. 

For the record, I removed the redundant check because `$insertId != 0` casts the value to an integer before comparison. This makes `$insertId !== false` redundant. The condition should therefore be `if($insertId != 0)`. But I think here is where I made a mistake/an assumption. If the returned value is an empty array (unlikely) or a non-empty non-numeric-string (possible) then the result of the comparison will be different. For this reason, I will add the `!= 0` back.

williamdes reviewed (approved):
Did not see anything alarming, did post some comments on minor things

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -1785,94 +1758,91 @@ public function getCurrentValueForDifferentTypes(
+            // (works also in non-strict mode)
+            if (isset($multiEditAutoIncrement, $multiEditAutoIncrement[$key])) {
                 $currentValue = 'NULL';
+            } else {
+                $currentValue = "''";
             }
```
The elses could be avoided by setting the value above the if

williamdes commented on libraries/classes/InsertEdit.php:
```diff
@@ -2130,15 +2092,14 @@ private function getHtmlForInsertEditFormColumn(
 
         $asIs = false;
-        if (! empty($repopulate) && ! empty($currentRow)) {
+        if ($repopulate && $currentRow) {
```
since currentRow is an array, could we narrow down what we are checking here and further ?

kamil-tekiela replied:
Hmm, yeah. What are we checking here? 😀

kamil-tekiela replied:
Ok, that's tough. `$repopulate` is a global. It has no defined structure. All we are really checking here is that both arrays are non-empty. I can only introduce `array_key_exists` here on `$repopulate`. 

I could remove the parameter altogether and replace it with the element that we need, but the problem is that I have absolutely no idea what this piece of code does. Unit tests don't cover this code and manual analysis of code flow leads me to madness... 

I will add `array_key_exists()` but if you have an idea on how to improve this piece of code, I would appreciate it.

williamdes replied:
Thank you for the `array_key_exists` ;)

kamil-tekiela committed:
Remove else statements

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Introduce array_key_exists() and a local variable

We also add phpdoc comment telling static analysis that the value should
be a string. This is what the current code expects already.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove object parameter type declaration

Upon further inspection I realize this was wrong. The unit tests rely
on the parameter type being an integer and so we cannot force it to be
only an object. I leave the type hint as object as array was probably
never the right type.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

mauriciofauth commented:
Merged. Thank you very much.
